Bug: [memory_checker] Failed to get container ID of <name> during config reload

### Is it platform specific

generic

### Importance or Severity

Medium

### Description of the bug

Memory-checker fails with `Failed to get container ID of <name>` during config reload, that is being caught by loganalyzer that causes multiple tests failure

### Steps to Reproduce

Flakey in nature
do a config reload and memory checker will occasionally emit the  error log to syslog: `2026 Jun 28 09:28:35.485290 bjw-can-7260-14 ERR memory_checker: [memory_checker] Failed to get container ID of 'gnmi'! Exiting ...`

### Actual Behavior and Expected Behavior

Actual Behavior: memory-checker is exiting with non-zero code
Expected Behavior: config reload case should be handled



### Relevant log output

```shell
>               pt_assert(False, failure_message)
E               Failed: Got matched syslog in processes "analyze_logs--<MultiAsicSonicHost bjw-can-7260-14>" exit code:"1"
E               match: 1
E               expected_match: 0
E               expected_missing_match: 0
E               
E               Match Messages:
E               2026 Jun 28 09:28:35.485290 bjw-can-7260-14 ERR memory_checker: [memory_checker] Failed to get container ID of 'gnmi'! Exiting ...
```

### Output of `show version`

```shell

```

### Attach files (if any)

_No response_

--- a/files/image_config/monit/memory_checker
+++ b/files/image_config/monit/memory_checker
@@ -83,6 +83,19 @@ def get_command_result(command):
 
     return command_stdout.strip()
 
+def exit_if_container_stopped(container_name):
+    """Exit gracefully if the container is no longer running.
+
+    Call before logging ERR to avoid false alarms when a container
+    disappears during config reload.
+    """
+    if container_name not in get_running_container_names():
+        syslog.syslog(syslog.LOG_INFO,
+                      "[memory_checker] Container '{}' stopped during memory check. Exiting gracefully."
+                      .format(container_name))
+        sys.exit(CONTAINER_NOT_RUNNING)
+
+
 def get_container_id(container_name):
     """Gets full container ID of the specified container
     Args:
@@ -102,17 +115,18 @@ def get_container_id(container_name):
             break
 
     if not container_id:
+        exit_if_container_stopped(container_name)
         syslog.syslog(syslog.LOG_ERR, ERROR_CONTAINER_ID_NOT_FOUND.format(container_name))
-
         sys.exit(INTERNAL_ERROR)
 
     return container_id
 
-def get_memory_usage(container_id):
+def get_memory_usage(container_id, container_name):
     """Reads the container's memory usage from the control group subsystem's file
     '/sys/fs/cgroup/memory/docker/<container_id>/memory.current'.
     Args:
         container_id: A string indicates the full ID of a container.
+        container_name: A string indicates the name of the container.
     Returns:
         A string indicates memory usage (Bytes) of a container.
     """
@@ -130,17 +144,20 @@ def get_memory_usage(container_id):
             else:
                 break
         except IOError:
+            exit_if_container_stopped(container_name)
             syslog.syslog(syslog.LOG_ERR, ERROR_CONTAINER_MEMORY_USAGE_NOT_FOUND.format(container_id))
             sys.exit(INTERNAL_ERROR)
 
+    exit_if_container_stopped(container_name)
     syslog.syslog(syslog.LOG_ERR, ERROR_CGROUP_MEMORY_USAGE_NOT_FOUND.format(docker_memory_usage_file_path, container_id))
     sys.exit(INTERNAL_ERROR)
 
-def get_inactive_cache_usage(container_id):
+def get_inactive_cache_usage(container_id, container_name):
     """Reads the container's cache usage from the field 'inactive_file' in control
     group subsystem's file '/sys/fs/cgroup/memory/docker/<container_id>/memory.stat'.
     Args:
         container_id: A string indicates the full ID of a container.
+        container_name: A string indicates the name of the container.
     Returns:
         cache_usage_in_bytes: A string indicates the cache usage (Bytes) of a container.
     """
@@ -150,6 +167,7 @@ def get_inactive_cache_usage(container_id):
 
     docker_memory_stat_file_path = CGROUP_DOCKER_MEMORY_DIR + container_id + ".scope/memory.stat"
     if not os.path.exists(docker_memory_stat_file_path):
+        exit_if_container_stopped(container_name)
         syslog.syslog(syslog.LOG_ERR, ERROR_CGROUP_MEMORY_STATS_NOT_FOUND.format(docker_memory_stat_file_path, container_id))
         sys.exit(INTERNAL_ERROR)
 
@@ -165,6 +183,7 @@ def get_inactive_cache_usage(container_id):
                         sys.exit(INTERNAL_ERROR)
                     break
     except IOError as err:
+        exit_if_container_stopped(container_name)
         syslog.syslog(syslog.LOG_ERR, ERROR_CONTAINER_CACHE_USAGE_NOT_FOUND.format(container_id))
         sys.exit(INTERNAL_ERROR)
 
@@ -199,11 +218,11 @@ def check_memory_usage(container_name, threshold_value):
     syslog.syslog(syslog.LOG_INFO, "[memory_checker] Container ID of '{}' is: '{}'."
                   .format(container_name, container_id))
 
-    memory_usage_in_bytes = get_memory_usage(container_id)
+    memory_usage_in_bytes = get_memory_usage(container_id, container_name)
     syslog.syslog(syslog.LOG_INFO, "[memory_checker] The memory usage of container '{}' is '{}' Bytes!"
                   .format(container_name, memory_usage_in_bytes))
 
-    cache_usage_in_bytes = get_inactive_cache_usage(container_id)
+    cache_usage_in_bytes = get_inactive_cache_usage(container_id, container_name)
     syslog.syslog(syslog.LOG_INFO, "[memory_checker] The cache usage of container '{}' is '{}' Bytes!"
                   .format(container_name, cache_usage_in_bytes))
 

--- a/files/image_config/monit/memory_checker
+++ b/files/image_config/monit/memory_checker
@@ -45,7 +45,7 @@ ERROR_CGROUP_MEMORY_STATS_NOT_FOUND = "[memory_checker] cgroup memory statistics
 ERROR_CGROUP_MEMORY_STATS_LINE_FORMAT = "[memory_checker] cgroup memory statistics file '{}' of container '{}' has invalid line format! Exiting ..."
 
 # Define common exit codes
-CONTAINER_NOT_RUNNING = 0
+NO_ERROR = 0
 INTERNAL_ERROR = 1
 INVALID_VALUE = 2
 EXCEED_THRESHOLD = 3
@@ -83,28 +83,32 @@ def get_command_result(command):
 
     return command_stdout.strip()
 
-class _ContainerGone:
-    """Sentinel returned by _try_get_container_id when the container no longer exists."""
+class _ContainerNotRunning:
+    """Sentinel returned by _try_get_container_id when the container is no longer running."""
     pass
 
 
-_CONTAINER_GONE = _ContainerGone()
+_CONTAINER_NOT_RUNNING = _ContainerNotRunning()
 
 
 def _try_get_container_id(container_name):
-    """Get the current container ID.
+    """Get the ID of the running container.
 
-    Returns the ID string, _CONTAINER_GONE if the container no longer
-    exists, or None on transient Docker API errors.
+    Returns the ID string if the container is running, _CONTAINER_NOT_RUNNING if
+    the container is not found or not running, or None on transient Docker
+    API errors.
     """
     try:
         docker_client = docker.DockerClient(base_url='unix://var/run/docker.sock')
         try:
-            return docker_client.containers.get(container_name).id
+            container = docker_client.containers.get(container_name)
+            if container.status != 'running':
+                return _CONTAINER_NOT_RUNNING
+            return container.id
         finally:
             docker_client.close()
     except docker.errors.NotFound:
-        return _CONTAINER_GONE
+        return _CONTAINER_NOT_RUNNING
     except (docker.errors.APIError, docker.errors.DockerException):
         return None
 
@@ -115,19 +119,13 @@ def exit_if_container_stopped(container_name, container_id=None):
     Call before logging ERR to avoid false alarms when a container
     disappears or restarts with a new ID during config reload.
     """
-    if container_name not in get_running_container_names():
+    current_id = _try_get_container_id(container_name)
+    if isinstance(current_id, _ContainerNotRunning) or (container_id and current_id and current_id != container_id):
+        reason = "stopped" if isinstance(current_id, _ContainerNotRunning) else "restarted"
         syslog.syslog(syslog.LOG_INFO,
-                      "[memory_checker] Container '{}' stopped during memory check. Exiting gracefully."
-                      .format(container_name))
-        sys.exit(CONTAINER_NOT_RUNNING)
-    if container_id:
-        current_id = _try_get_container_id(container_name)
-        if isinstance(current_id, _ContainerGone) or (current_id and current_id != container_id):
-            reason = "removed" if isinstance(current_id, _ContainerGone) else "restarted"
-            syslog.syslog(syslog.LOG_INFO,
-                          "[memory_checker] Container '{}' {} during memory check. Exiting gracefully."
-                          .format(container_name, reason))
-            sys.exit(CONTAINER_NOT_RUNNING)
+                      "[memory_checker] Container '{}' {} during memory check. Exiting gracefully."
+                      .format(container_name, reason))
+        sys.exit(NO_ERROR)
 
 
 def get_container_id(container_name):
@@ -301,9 +299,12 @@ def get_running_container_names():
     """
     try:
         docker_client = docker.DockerClient(base_url='unix://var/run/docker.sock')
-        running_container_list = docker_client.containers.list(filters={"status": "running"},
-                                                               ignore_removed=True)
-        running_container_names = [ container.name for container in running_container_list ]
+        try:
+            running_container_list = docker_client.containers.list(filters={"status": "running"},
+                                                                   ignore_removed=True)
+            running_container_names = [ container.name for container in running_container_list ]
+        finally:
+            docker_client.close()
     except (docker.errors.APIError, docker.errors.DockerException) as err:
         if not is_service_active("docker"):
             syslog.syslog(syslog.LOG_INFO,
@@ -332,7 +333,7 @@ def main():
         syslog.syslog(syslog.LOG_INFO,
                       "[memory_checker] Exits without checking memory usage of container '{}' since docker daemon is not running!"
                       .format(args.container_name))
-        sys.exit(CONTAINER_NOT_RUNNING)
+        sys.exit(NO_ERROR)
 
     running_container_names = get_running_container_names()
     if args.container_name in running_container_names:

--- a/files/image_config/monit/memory_checker
+++ b/files/image_config/monit/memory_checker
@@ -83,17 +83,36 @@ def get_command_result(command):
 
     return command_stdout.strip()
 
-def exit_if_container_stopped(container_name):
-    """Exit gracefully if the container is no longer running.
+def _try_get_container_id(container_name):
+    """Get the current container ID, or None if not found."""
+    try:
+        docker_client = docker.DockerClient(base_url='unix://var/run/docker.sock')
+        try:
+            return docker_client.containers.get(container_name).id
+        finally:
+            docker_client.close()
+    except (docker.errors.NotFound, docker.errors.APIError, docker.errors.DockerException):
+        return None
+
+
+def exit_if_container_stopped(container_name, container_id=None):
+    """Exit gracefully if the container stopped or restarted.
 
     Call before logging ERR to avoid false alarms when a container
-    disappears during config reload.
+    disappears or restarts with a new ID during config reload.
     """
     if container_name not in get_running_container_names():
         syslog.syslog(syslog.LOG_INFO,
                       "[memory_checker] Container '{}' stopped during memory check. Exiting gracefully."
                       .format(container_name))
         sys.exit(CONTAINER_NOT_RUNNING)
+    if container_id:
+        current_id = _try_get_container_id(container_name)
+        if current_id and current_id != container_id:
+            syslog.syslog(syslog.LOG_INFO,
+                          "[memory_checker] Container '{}' restarted with a new ID during memory check. Exiting gracefully."
+                          .format(container_name))
+            sys.exit(CONTAINER_NOT_RUNNING)
 
 
 def get_container_id(container_name):
@@ -144,11 +163,11 @@ def get_memory_usage(container_id, container_name):
             else:
                 break
         except IOError:
-            exit_if_container_stopped(container_name)
+            exit_if_container_stopped(container_name, container_id)
             syslog.syslog(syslog.LOG_ERR, ERROR_CONTAINER_MEMORY_USAGE_NOT_FOUND.format(container_id))
             sys.exit(INTERNAL_ERROR)
 
-    exit_if_container_stopped(container_name)
+    exit_if_container_stopped(container_name, container_id)
     syslog.syslog(syslog.LOG_ERR, ERROR_CGROUP_MEMORY_USAGE_NOT_FOUND.format(docker_memory_usage_file_path, container_id))
     sys.exit(INTERNAL_ERROR)
 
@@ -167,7 +186,7 @@ def get_inactive_cache_usage(container_id, container_name):
 
     docker_memory_stat_file_path = CGROUP_DOCKER_MEMORY_DIR + container_id + ".scope/memory.stat"
     if not os.path.exists(docker_memory_stat_file_path):
-        exit_if_container_stopped(container_name)
+        exit_if_container_stopped(container_name, container_id)
         syslog.syslog(syslog.LOG_ERR, ERROR_CGROUP_MEMORY_STATS_NOT_FOUND.format(docker_memory_stat_file_path, container_id))
         sys.exit(INTERNAL_ERROR)
 
@@ -183,7 +202,7 @@ def get_inactive_cache_usage(container_id, container_name):
                         sys.exit(INTERNAL_ERROR)
                     break
     except IOError as err:
-        exit_if_container_stopped(container_name)
+        exit_if_container_stopped(container_name, container_id)
         syslog.syslog(syslog.LOG_ERR, ERROR_CONTAINER_CACHE_USAGE_NOT_FOUND.format(container_id))
         sys.exit(INTERNAL_ERROR)
 
@@ -267,8 +286,7 @@ def get_running_container_names():
     """
     try:
         docker_client = docker.DockerClient(base_url='unix://var/run/docker.sock')
-        running_container_list = docker_client.containers.list(filters={"status": "running"},
-                                                               ignore_removed=True)
+        running_container_list = docker_client.containers.list(filters={"status": "running"})
         running_container_names = [ container.name for container in running_container_list ]
     except (docker.errors.APIError, docker.errors.DockerException) as err:
         if not is_service_active("docker"):

--- a/files/image_config/monit/memory_checker
+++ b/files/image_config/monit/memory_checker
@@ -83,15 +83,29 @@ def get_command_result(command):
 
     return command_stdout.strip()
 
+class _ContainerGone:
+    """Sentinel returned by _try_get_container_id when the container no longer exists."""
+    pass
+
+
+_CONTAINER_GONE = _ContainerGone()
+
+
 def _try_get_container_id(container_name):
-    """Get the current container ID, or None if not found."""
+    """Get the current container ID.
+
+    Returns the ID string, _CONTAINER_GONE if the container no longer
+    exists, or None on transient Docker API errors.
+    """
     try:
         docker_client = docker.DockerClient(base_url='unix://var/run/docker.sock')
         try:
             return docker_client.containers.get(container_name).id
         finally:
             docker_client.close()
-    except (docker.errors.NotFound, docker.errors.APIError, docker.errors.DockerException):
+    except docker.errors.NotFound:
+        return _CONTAINER_GONE
+    except (docker.errors.APIError, docker.errors.DockerException):
         return None
 
 
@@ -108,10 +122,11 @@ def exit_if_container_stopped(container_name, container_id=None):
         sys.exit(CONTAINER_NOT_RUNNING)
     if container_id:
         current_id = _try_get_container_id(container_name)
-        if current_id and current_id != container_id:
+        if isinstance(current_id, _ContainerGone) or (current_id and current_id != container_id):
+            reason = "removed" if isinstance(current_id, _ContainerGone) else "restarted"
             syslog.syslog(syslog.LOG_INFO,
-                          "[memory_checker] Container '{}' restarted with a new ID during memory check. Exiting gracefully."
-                          .format(container_name))
+                          "[memory_checker] Container '{}' {} during memory check. Exiting gracefully."
+                          .format(container_name, reason))
             sys.exit(CONTAINER_NOT_RUNNING)
 
 
@@ -286,7 +301,8 @@ def get_running_container_names():
     """
     try:
         docker_client = docker.DockerClient(base_url='unix://var/run/docker.sock')
-        running_container_list = docker_client.containers.list(filters={"status": "running"})
+        running_container_list = docker_client.containers.list(filters={"status": "running"},
+                                                               ignore_removed=True)
         running_container_names = [ container.name for container in running_container_list ]
     except (docker.errors.APIError, docker.errors.DockerException) as err:
         if not is_service_active("docker"):

--- a/files/image_config/monit/tests/test_memory_checker.py
+++ b/files/image_config/monit/tests/test_memory_checker.py
@@ -25,8 +25,9 @@ class TestMemoryChecker(unittest.TestCase):
         mock_popen.return_value.communicate.assert_called_with()
         self.assertEqual(mock_popen.return_value.returncode, returncode)
 
+    @patch('memory_checker.exit_if_container_stopped')
     @patch('memory_checker.get_command_result')
-    def test_get_container_id(self, mock_get_command_result):
+    def test_get_container_id(self, mock_get_command_result, mock_exit_check):
         container_name = 'your_container'
         command = ['docker', 'ps', '--no-trunc', '--filter', 'name=your_container']
         mock_get_command_result.return_value = ''
@@ -36,26 +37,43 @@ class TestMemoryChecker(unittest.TestCase):
         self.assertEqual(cm.exception.code, 1)
         mock_get_command_result.assert_called_once_with(command)
 
+    @patch('memory_checker.exit_if_container_stopped')
     @patch('memory_checker.open', side_effect=FileNotFoundError)
-    def test_get_memory_usage(self, mock_open):
+    def test_get_memory_usage(self, mock_open, mock_exit_check):
         container_id = 'your_container_id'
+        container_name = 'your_container'
         with self.assertRaises(SystemExit) as cm:
-            memory_checker.get_memory_usage(container_id)
+            memory_checker.get_memory_usage(container_id, container_name)
         self.assertEqual(cm.exception.code, 1)
 
     @patch('memory_checker.open', side_effect=FileNotFoundError)
     def test_get_memory_usage_invalid(self, mock_open):
         container_id = '../..'
+        container_name = 'your_container'
         with self.assertRaises(SystemExit) as cm:
-            memory_checker.get_memory_usage(container_id)
+            memory_checker.get_memory_usage(container_id, container_name)
         self.assertEqual(cm.exception.code, 1)
 
+    @patch('memory_checker.exit_if_container_stopped')
     @patch('builtins.open', side_effect=FileNotFoundError)
-    def test_get_inactive_cache_usage(self, mock_open):
+    def test_get_inactive_cache_usage(self, mock_open, mock_exit_check):
         container_id = 'your_container_id'
+        container_name = 'your_container'
         with self.assertRaises(SystemExit) as cm:
-            memory_checker.get_inactive_cache_usage(container_id)
+            memory_checker.get_inactive_cache_usage(container_id, container_name)
         self.assertEqual(cm.exception.code, 1)
+
+    @patch('memory_checker.get_running_container_names', return_value=[])
+    def test_exit_if_container_stopped(self, mock_running):
+        """Container not running - should exit gracefully."""
+        with self.assertRaises(SystemExit) as cm:
+            memory_checker.exit_if_container_stopped('gnmi')
+        self.assertEqual(cm.exception.code, 0)
+
+    @patch('memory_checker.get_running_container_names', return_value=['gnmi'])
+    def test_exit_if_container_stopped_still_running(self, mock_running):
+        """Container still running - should not exit."""
+        memory_checker.exit_if_container_stopped('gnmi')
 
     @patch('syslog.syslog')
     @patch('memory_checker.get_container_id')
@@ -75,7 +93,7 @@ class TestMemoryChecker(unittest.TestCase):
             memory_checker.check_memory_usage(container_name, threshold_value)
 
         self.assertEqual(cm.exception.code, 3)
-        mock_get_memory_usage.assert_called_once_with(container_name)
+        mock_get_memory_usage.assert_called_once_with(container_id, container_name)
 
 if __name__ == '__main__':
     unittest.main()

--- a/files/image_config/monit/tests/test_memory_checker.py
+++ b/files/image_config/monit/tests/test_memory_checker.py
@@ -89,6 +89,20 @@ class TestMemoryChecker(unittest.TestCase):
         """Container running with same ID - should not exit."""
         memory_checker.exit_if_container_stopped('gnmi', container_id='same_id_123')
 
+    @patch('memory_checker._try_get_container_id', return_value=memory_checker._CONTAINER_GONE)
+    @patch('memory_checker.get_running_container_names', return_value=['gnmi'])
+    def test_exit_if_container_removed_between_checks(self, mock_running, mock_get_id):
+        """Container in running list but gone by ID lookup - should exit gracefully."""
+        with self.assertRaises(SystemExit) as cm:
+            memory_checker.exit_if_container_stopped('gnmi', container_id='old_id_123')
+        self.assertEqual(cm.exception.code, 0)
+
+    @patch('memory_checker._try_get_container_id', return_value=None)
+    @patch('memory_checker.get_running_container_names', return_value=['gnmi'])
+    def test_exit_if_docker_api_error_does_not_exit(self, mock_running, mock_get_id):
+        """Docker API error (None) should not cause graceful exit - let original error path handle it."""
+        memory_checker.exit_if_container_stopped('gnmi', container_id='old_id_123')
+
     @patch('syslog.syslog')
     @patch('memory_checker.get_container_id')
     @patch('memory_checker.get_memory_usage')

--- a/files/image_config/monit/tests/test_memory_checker.py
+++ b/files/image_config/monit/tests/test_memory_checker.py
@@ -72,8 +72,22 @@ class TestMemoryChecker(unittest.TestCase):
 
     @patch('memory_checker.get_running_container_names', return_value=['gnmi'])
     def test_exit_if_container_stopped_still_running(self, mock_running):
-        """Container still running - should not exit."""
+        """Container still running, no container_id check - should not exit."""
         memory_checker.exit_if_container_stopped('gnmi')
+
+    @patch('memory_checker._try_get_container_id', return_value='new_id_456')
+    @patch('memory_checker.get_running_container_names', return_value=['gnmi'])
+    def test_exit_if_container_restarted(self, mock_running, mock_get_id):
+        """Container running with different ID - should exit gracefully."""
+        with self.assertRaises(SystemExit) as cm:
+            memory_checker.exit_if_container_stopped('gnmi', container_id='old_id_123')
+        self.assertEqual(cm.exception.code, 0)
+
+    @patch('memory_checker._try_get_container_id', return_value='same_id_123')
+    @patch('memory_checker.get_running_container_names', return_value=['gnmi'])
+    def test_exit_if_container_same_id(self, mock_running, mock_get_id):
+        """Container running with same ID - should not exit."""
+        memory_checker.exit_if_container_stopped('gnmi', container_id='same_id_123')
 
     @patch('syslog.syslog')
     @patch('memory_checker.get_container_id')

--- a/files/image_config/monit/tests/test_memory_checker.py
+++ b/files/image_config/monit/tests/test_memory_checker.py
@@ -63,45 +63,81 @@ class TestMemoryChecker(unittest.TestCase):
             memory_checker.get_inactive_cache_usage(container_id, container_name)
         self.assertEqual(cm.exception.code, 1)
 
-    @patch('memory_checker.get_running_container_names', return_value=[])
-    def test_exit_if_container_stopped(self, mock_running):
+    @patch('memory_checker._try_get_container_id', return_value=memory_checker._CONTAINER_NOT_RUNNING)
+    def test_exit_if_container_stopped(self, mock_get_id):
         """Container not running - should exit gracefully."""
         with self.assertRaises(SystemExit) as cm:
             memory_checker.exit_if_container_stopped('gnmi')
         self.assertEqual(cm.exception.code, 0)
 
-    @patch('memory_checker.get_running_container_names', return_value=['gnmi'])
-    def test_exit_if_container_stopped_still_running(self, mock_running):
-        """Container still running, no container_id check - should not exit."""
+    @patch('memory_checker._try_get_container_id', return_value='current_id_123')
+    def test_exit_if_container_still_running(self, mock_get_id):
+        """Container running, no container_id check - should not exit."""
         memory_checker.exit_if_container_stopped('gnmi')
 
     @patch('memory_checker._try_get_container_id', return_value='new_id_456')
-    @patch('memory_checker.get_running_container_names', return_value=['gnmi'])
-    def test_exit_if_container_restarted(self, mock_running, mock_get_id):
+    def test_exit_if_container_restarted(self, mock_get_id):
         """Container running with different ID - should exit gracefully."""
         with self.assertRaises(SystemExit) as cm:
             memory_checker.exit_if_container_stopped('gnmi', container_id='old_id_123')
         self.assertEqual(cm.exception.code, 0)
 
     @patch('memory_checker._try_get_container_id', return_value='same_id_123')
-    @patch('memory_checker.get_running_container_names', return_value=['gnmi'])
-    def test_exit_if_container_same_id(self, mock_running, mock_get_id):
+    def test_exit_if_container_same_id(self, mock_get_id):
         """Container running with same ID - should not exit."""
         memory_checker.exit_if_container_stopped('gnmi', container_id='same_id_123')
 
-    @patch('memory_checker._try_get_container_id', return_value=memory_checker._CONTAINER_GONE)
-    @patch('memory_checker.get_running_container_names', return_value=['gnmi'])
-    def test_exit_if_container_removed_between_checks(self, mock_running, mock_get_id):
-        """Container in running list but gone by ID lookup - should exit gracefully."""
+    @patch('memory_checker._try_get_container_id', return_value=memory_checker._CONTAINER_NOT_RUNNING)
+    def test_exit_if_container_removed(self, mock_get_id):
+        """Container gone by ID lookup - should exit gracefully."""
         with self.assertRaises(SystemExit) as cm:
             memory_checker.exit_if_container_stopped('gnmi', container_id='old_id_123')
         self.assertEqual(cm.exception.code, 0)
 
     @patch('memory_checker._try_get_container_id', return_value=None)
-    @patch('memory_checker.get_running_container_names', return_value=['gnmi'])
-    def test_exit_if_docker_api_error_does_not_exit(self, mock_running, mock_get_id):
+    def test_exit_if_docker_api_error_does_not_exit(self, mock_get_id):
         """Docker API error (None) should not cause graceful exit - let original error path handle it."""
         memory_checker.exit_if_container_stopped('gnmi', container_id='old_id_123')
+
+    @patch('memory_checker.docker')
+    def test_try_get_container_id_running(self, mock_docker):
+        """Container running - should return its ID."""
+        mock_container = MagicMock()
+        mock_container.id = 'abc123'
+        mock_container.status = 'running'
+        mock_docker.DockerClient.return_value.containers.get.return_value = mock_container
+        result = memory_checker._try_get_container_id('gnmi')
+        self.assertEqual(result, 'abc123')
+
+    @patch('memory_checker.docker')
+    def test_try_get_container_id_stopped(self, mock_docker):
+        """Container exists but stopped - should return _CONTAINER_NOT_RUNNING."""
+        mock_container = MagicMock()
+        mock_container.id = 'abc123'
+        mock_container.status = 'exited'
+        mock_docker.DockerClient.return_value.containers.get.return_value = mock_container
+        result = memory_checker._try_get_container_id('gnmi')
+        self.assertIsInstance(result, memory_checker._ContainerNotRunning)
+
+    @patch('memory_checker.docker')
+    def test_try_get_container_id_not_found(self, mock_docker):
+        """Container not found - should return _CONTAINER_NOT_RUNNING."""
+        import docker as real_docker
+        mock_docker.errors.NotFound = real_docker.errors.NotFound
+        mock_docker.DockerClient.return_value.containers.get.side_effect = real_docker.errors.NotFound('not found')
+        result = memory_checker._try_get_container_id('gnmi')
+        self.assertIsInstance(result, memory_checker._ContainerNotRunning)
+
+    @patch('memory_checker.docker')
+    def test_try_get_container_id_api_error(self, mock_docker):
+        """Docker API error - should return None."""
+        import docker as real_docker
+        mock_docker.errors.NotFound = real_docker.errors.NotFound
+        mock_docker.errors.APIError = real_docker.errors.APIError
+        mock_docker.errors.DockerException = real_docker.errors.DockerException
+        mock_docker.DockerClient.return_value.containers.get.side_effect = real_docker.errors.APIError('error')
+        result = memory_checker._try_get_container_id('gnmi')
+        self.assertIsNone(result)
 
     @patch('syslog.syslog')
     @patch('memory_checker.get_container_id')

